include/git2/common.h

```diff
@@ -205,7 +205,9 @@ typedef enum {
 	GIT_OPT_GET_PACK_MAX_OBJECTS,
 	GIT_OPT_SET_PACK_MAX_OBJECTS,
 	GIT_OPT_DISABLE_PACK_KEEP_FILE_CHECKS,
-	GIT_OPT_ENABLE_HTTP_EXPECT_CONTINUE
+	GIT_OPT_ENABLE_HTTP_EXPECT_CONTINUE,
+	GIT_OPT_SET_ODB_PACKED_PRIORITY,
+	GIT_OPT_SET_ODB_LOOSE_PRIORITY
```

Comment:
I'm sorry to be late to the game and propose a different angle, but I don't think having this as a global option is a sensible thing to do. Talking about this in a matter of priorities feels really indirect to me and not that great from a usability point of perspective. That's not your fault but ours, but I'd like to provide a better interface, especially as we will have to do the same for reference databases at some point in time.

Our existing implementation doesn't really allow for some special usecases.

- Somebody who never wants any loose objects at all may want to use a packed ODB implementation only, or vice versa.
- It's hard to couple custom backends with given ones.
- We do not easily allow for creation of repos with a custom set of ODB/refbd implementations.

So in the future, I have the intent to extend ODB/refdbs with some kind of constructor that allows it to say "Here's a new repo that wasn't initialized yet. Do your thing and set up data structures you expect the gitdir to have." That will require some refactoring, as we've got that logic all over the place right now for refdbs, and I expect it's the same for the ODB. But most importantly, this cannot be solved via global priorities. While it would work for our builtin ones, it wouldn't for external ones.

So what do I expect it to look like in the future? I'm not quite sure yet, to be honest. But I think that the safest way to approach this right now is to add a field to `git_repository_open_options` that allows users to pass in an array of object/reference database constructors. A similar thing could be done for `git_repository_new`: pass in a set of constructors that, given a gitdir path, set up their data structures and return a new refdb/odb. Priorities would then be derived from the index of that array.

What do you think?

Reply:
Do you thinking having repository open taken in a vector of backends meet the requirement? The contents and order in the vector will define the backends as requested by the caller. It allows adding user implemented object backends too.

Since we want to support user implemented backends too, it might be best to avoid `git_repository_open_options`. IIUC, any option should be defined in libgit2 and this will prevent treating user implemented object backends as first class citizens.

In short, I am proposing a new function to take in a vector of backends and register it as object backend in the same order as defined in the vector. If there is an alternates file, register the same backends as alternates.

Looking forward to hearing your feedback.

Reply:
Taking in an array/vector of backends would work. However, we could lose the ability to lazy instantiate the backends like it happens today for bare repositories.

If a user just wants to control the order of default backend and make it behave like the git binary, we will be forced to instantiate the loose and pack backends before opening the repository.